dids/src/main/kotlin/web5/sdk/dids/methods/dht/DidDht.kt

```diff
@@ -45,393 +58,492 @@ public class DidDhtConfiguration internal constructor(
-      val publicKeyJwk = identityKey.toPublicJWK()
-      val publicKeyBytes = Crypto.publicKeyToBytes(publicKeyJwk)
-      val zBase32Encoded = ZBase32.encode(publicKeyBytes)
-      return "did:dht:$zBase32Encoded"
+    // build DID Document
+    val didDocument =
+      DIDDocument.builder()
+        .id(URI(id))
+        .verificationMethods(verificationMethods)
+        .services(services)
+        .assertionMethodVerificationMethods(relationshipsMap[PublicKeyPurpose.ASSERTION_METHOD])
+        .authenticationVerificationMethods(relationshipsMap[PublicKeyPurpose.AUTHENTICATION])
+        .keyAgreementVerificationMethods(relationshipsMap[PublicKeyPurpose.KEY_AGREEMENT])
+        .capabilityDelegationVerificationMethods(relationshipsMap[PublicKeyPurpose.CAPABILITY_DELEGATION])
+        .capabilityInvocationVerificationMethods(relationshipsMap[PublicKeyPurpose.CAPABILITY_INVOCATION])
+        .build()
+
+    // publish to DHT if requested
+    if (opts.publish) {
+      publish(keyManager, didDocument)
     }
 
-    /**
-     * Checks whether a given DID identifier conforms to the "did:dht" method. This checks that the DID starts with
-     * "did:dht" and that the suffix is a valid z-base-32 encoded public key.
-     *
-     * @param did The DID to check.
-     * @throws IllegalArgumentException
-     * @throws ParserException
-     */
-    public fun validate(did: String) {
-      val parsedDid = DID.fromString(did)
-      require(parsedDid.methodName == "dht") { "expected method to be dht" }
-
-      val decodedId = ZBase32.decode(parsedDid.methodSpecificId)
-      require(decodedId.size == 32) { "expected size of decoded identifier to be 32" }
+    return DidDht(id, keyManager, didDocument, this)
+  }
+
+  /**
+   * Resolves a "did:dht" DID into a [DidResolutionResult], which contains the DID Document and possible related
+   * metadata.
+   *
+   * This implementation talks to a DID DHT gateway to retrieve the DID Document, which in turn takes the z-base-32
+   * encoded identifier public key and uses it to retrieve the DID Document from the DHT. Next, the Pkarr response is
+   * parsed, and used to reconstruct the DID Document.
+   *
+   * @param did The "did:dht" DID that needs to be resolved.
+   * @return A [DidResolutionResult] instance containing the DID Document and related context, including types
+   * as part of the [DidDocumentMetadata], if available.
+   * @throws IllegalArgumentException if the provided DID does not conform to the "did:dht" method.
+   */
+  override fun resolve(did: String, options: ResolveDidOptions?): DidResolutionResult {
+    validate(did)
+    val getId = DidDht.suffix(did)
+    val bep44Message = dht.pkarrGet(getId)
+    val dnsPacket = Dht.parseBep44GetResponse(bep44Message)
+    fromDnsPacket(did, dnsPacket).let { (didDocument, types) ->
+      return DidResolutionResult(didDocument = didDocument, didDocumentMetadata = DidDocumentMetadata(types = types))
     }
+  }
```

Comment:
I imagine `resolve` would _always_ return a `DidResolutionResult`. Errors are caught and reflected in [`didResolutionMetadata`](https://www.w3.org/TR/did-core/#did-resolution-metadata)

Reply:
this method distinguishes between errors before the network call (argument exceptions) and after (resolution results) -- this mirrors the functionality we already have in both did key and did ion implementations